src: optimizing watchdog performance

Gefix opened the pull request:
Performance of vm.Script's runInContext for small and/or quick scripts can be much slower when the optional execution timeout parameter is set - over 30 times slower in edge cases.

The root issue is the constant creation and destruction of threads. Each timeout is enforced with a parallel uv loop that contains a timer and for this a new uv thread is created and later destroyed.

For applications that utilize the vm for sandboxing it is often the case that the optional timeout will be frequently, if not always, set.

This is a small rework of the watchdog that will transform the uv loop into a single parallel thread for each execution thread.

Instead of spawning a thread for each watchdog instance a thread-local WatchdogService is used to execute the timer logic for all timeouts from the given execution thread. The WatchdogService is lazy-initialized the first time a watchdog is requested from an execution thread.

The uv_loop and uv_timers used by the Watchdog have been replaced by a multimap-based priority queue of WatchdogTimer objects. Timekeeping is done using uv_hrtime,

The WatchdogService::Run loop sleeps until the closest timeout is about to expire, or forever if there are no timers set. A conditional wait is used to notify the run loop of a new timer only if its expiration time is shorter than the current time the run loop is about to awaken.

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

bnoordhuis reviewed (commented):
Only lightly reviewed but it looks to be doing something that libuv explicitly doesn't support.

Gefix commented:
@bnoordhuis, thank you for looking into this PR! I have updated the changes to blend in with the style and also fixed some of the linting and formatting errors.

At this point, I would like to acknowledge that the chances of this proposal making it to the main branch in its current state are slim to none.

It's not just the mutex over the loop; there's also a thread sleep and an infinite while loop.

The task is quite niche, and this submission serves more as a conversation starter and an attempt to gather suggestions on how to better approach this from the perspective of the Node.js project.

This is merely the crossing of the first milestone – making it work without crashing. This idea has been brewing for a few years now, and the original patch has been used in production to speed up sandboxed expression evaluations.

The final goal is a clear one: achieving a substantial speedup in the following benchmark for the "with timeout" case:

`const vm = require('vm');

const script = new vm.Script('x += 1');
const iterations = 100_000;

const benchmark = (testFn) => {
  const start = process.hrtime.bigint();
  testFn();
  const end = process.hrtime.bigint();
  console.log(`Elapsed time for test ${testFn.name.padEnd(30)} is: ${(end - start) / 1000000n} ms`)
};

const runInNewContextWithoutTimeout = () => {
  const context = { x: 0 };
  for (let i = 0; i < iterations; i++) {
    script.runInNewContext(context);
  }
};

const runInNewContextWithTimeout = () => {
  const context = { x: 0 };
  for (let i = 0; i < iterations; i++) {
    script.runInNewContext(context, { timeout: 100 });
  }
}

benchmark(runInNewContextWithoutTimeout);
benchmark(runInNewContextWithTimeout);
`

bnoordhuis commented:
An approach that could work (although you should measure if it's actually faster) is as follows:

1. start a watchdog thread on first use
2. create a `uv_async_t` and a mutex-protected queue for message passing
3. other threads post "watch me" messages before running JS code
4. other threads post "done" messages after running JS code
5. watchdog thread calls `isolate->TerminateExecution()` if "done" message doesn't arrive on time

Technically speaking you don't even need to create a libuv event loop in step 2. A mutex and a condition variable with a timed wait are sufficient and likely has lower latency.

> the original patch has been used in production to speed up sandboxed expression evaluations

I'm curious, what are you doing exactly / who do you work for/with?

Gefix commented:
I was trying to preserve the original code as much as possible, with the uv_loop and the timer callbacks. However, you're right - it becomes somewhat of a hindrance in this case. Dropping the uv_loop altogether simplifies the implementation and reduces the amount of code needed, I'm not certain if using a multimap is the optimal solution for the priority queue here, but it seems to do the job.

With the introduction of the conditional wait, we've also eliminated the minimum timeout limit. The performance is now even more responsive with a 1ms timeout compared to the main branch. I'm curious about your thoughts on the default value used for the "forever sleep" (-1ull). Is it safe for all platforms, or should we opt for a reasonably large yet finite duration, such as 2^32 or something similar?

I'm also not entirely fond of the introduced thread_local services, but I'm not well-versed in the Node source code to determine if and where we can properly manage all the (de)initialization.

Regarding what we are doing with this and why we needed to patch it - Pliant's workflow engine relies heavily on time-boxed micro-evaluations.

bnoordhuis commented on src/node_watchdog.cc:
```diff
@@ -39,66 +39,107 @@ using v8::Local;
+    uint64_t sleep_nanos = ws->sleep_until_hrtime_ - hrtime;
+    uv_cond_timedwait(&ws->cond_, &ws->mutex_, sleep_nanos);
+    uv_mutex_unlock(&ws->mutex_);
+  }
```
The logic here doesn't look quite right to me. It should enter the loop holding the lock and not drop it until it breaks out of the loop again.

Gefix replied:
Quite right you are.

bnoordhuis reviewed (commented):
Only lightly reviewed because it's way past midnight here. :-)

Gefix committed:
src: optimizing watchdog performance

Performance of vm.Script's runInContext for small and/or quick scripts
can be much slower when the optional execution timeout parameter is set
 - over 30 times slower in edge cases.

The root issue is the constant creation and destruction of threads.
Each timeout is enforced with a parallel uv loop that contains a timer
and for this a new uv thread is created and later destroyed.

For applications that utilize the vm for sandboxing it is often the
case that the optional timeout will be frequently, if not always, set.

This is a small rework of the watchdog that will transform the uv loop
into a single parallel thread for each execution thread.

Instead of spawning a thread for each watchdog instance a thread-local
WatchdogService is used to execute the timer logic for all timeouts
from the given execution thread.
The WatchdogService is lazy-initialized the first time a watchdog is
requested from an execution thread.

The uv_loop and uv_timers used by the Watchdog have been replaced by a
multimap-based priority queue of WatchdogTimer objects.
Timekeeping is done using uv_hrtime.

The WatchdogService::Run loop sleeps until the closest timeout is about
to expire, or forever if there are no timers set. A conditional wait is
used to notify the run loop of a new timer only if its expiration time
is shorter than the current time the run loop is about to awaken.

bnoordhuis commented on src/node_watchdog.h:
```diff
@@ -42,24 +42,40 @@ enum class SignalPropagation {
-  uv_async_t async_;
-  uv_timer_t timer_;
-  bool* timed_out_;
+  WatchdogTimer* watchdog_timer_;
```
Node generally tries to use smart pointers when possible, i.e.:
```suggestion
  std::unique_ptr<WatchdogTimer> watchdog_timer_;
```

bnoordhuis replied:
A separate commit is fine. We'll squash it on merge anyway.

Gefix replied:
While we're on the WatchdogTimer struct - do you think it will be preferable if I reverse its members back into the main Watchdog class and just pass it's this around as it was done previously?

bnoordhuis reviewed (approved):
LGTM with a suggestion.

@addaleax you may also want to take a look.

Gefix committed:
src: switching to shared_ptr for the WatchdogTimer in Watchdog and WatchdogService

Gefix commented:
Shall I move this to Ready for review?

Gefix commented:
Please let me know if there's anything I can do to help move this forward. Thanks!

jasnell commented:
Needs to be rebased with new successful CI run. Removed the author ready label
